elastic/apm-integration-testing
# [Feature] Dyno

## What does this PR do?

This is the primary PR which introduces the capability to dynamically generate and control load for Opbeans as well as the capability to constrain the Opbeans applications in various ways, such as memory, CPU, and network behavior.

## Why is it important?

One major challenge in developing for APM comes from the desire to develop features against real data, generated from applications which are actually under duress. 

Additionally, it is challenging for PMs and even marketing/sales people to demonstrate the full feature set of APM due to the challenges of showing the relationship between a failing an application and the corresponding signals of that failure which may be emitted by the Elastic's APM solution.

This provides a solution by allowing real applications to have real load applied, and then to introduce real failures in any combination of ways in order to see how APM responds to those failures.

## What is included

The following new services are deployed when the `--dyno` flag is given to the test suite.

* The Opbeans load generator running in "web mode". For more information on this, please see [this PR](https://github.com/elastic/opbeans-loadgen/pull/17).
* A [Redis server](https://redis.io/), which is used to collate real-time load statistics and pass them along to the web browser.
* A [Toxiproxy server](https://github.com/Shopify/toxiproxy/releases) which will be configured to insert itself in between the network links for all configured services.
* The new Dyno management server, which serves the web application both by delivering the static assets (HTML/JS/CSS) and acting as a REST API for web interface.

## Screenshots

<img width="1649" alt="Capture d’écran 2021-01-12 à 10 51 21" src="https://user-images.githubusercontent.com/111616/104298344-c1aead80-54bb-11eb-9dd4-ff867c1f5155.png">

## Testing this PR

Please follow these steps to play around with this PR:

1. Check out PR locally
2. Ensure you have Python 3. (Python 2 is not supported.)
3. Start the integration test suite with the following command from the root of the project: `./scripts/compose.py start --with-opbeans-python --with-opbeans-python01 --dyno master`
4. Open the following in your browser: `http://localhost:9000`. This should give you the management interface shown above.
5. Open Kibana in a separate tab or window: `http://localhost:5601`. Navigate to the APM application and select the monitored Python application.
6. Start a load test by ticking the `Load test` box for one or both of the Python Opbeans applications in the Dyno inteface (from step 4).
7. Observe load being applied to the application in Kibana.
8. Play with the various sliders to adjust load and constraint characteristics from the Dyno interface.

## Reviewing this PR
The only JavaScript file which should be reviewed is called `dyno.js`. The others are bundled from 3rd-party libraries.

## Known bugs and caveats

* No JavaScript tests (yet). [Jest](https://jestjs.io/) is ready to go but as of yet, no tests have been written.
* Occasional crash of proxy servers during updates to network properties. If this happens, you will see all traffic drop to zero.  To remedy the situation, you need to restart the `toxi` container: `docker-compose restart toxi`. I am exploring a few ways to mitigate this problem, but it's still happening occasionally.
* Possible memory leak on the Dyno web interface. If you leave the Dyno management page open for many hours or days while a load test is running, you may experience sluggishness across all your tabs in your browser. Reloading the page will fix the problem.
* Not all Opbeans are supported. Python definitely works and Java should work. Other Opbeans either do not work at all or have not been tested. Please do not attempt to use them with Dyno.
* The REST API is designed for use directly in tests or with one's scripting language of choice in order to create tests which utilize the Dyno framework. However, it is not yet fully documented.  
* The CPU constraint, at times, may show little to no effect on application performance. Additional tuning will be required to find a suitable range for most scenarios.

## Related issues
Refs https://github.com/elastic/opbeans-loadgen/pull/17
